optuna/integration/pytorch_lightning.py

```diff
@@ -56,3 +59,87 @@ def on_validation_end(self, trainer: Trainer, pl_module: LightningModule) -> Non
+            ``val_acc``. The metrics are obtained from the returned dictionaries from e.g.
+            ``pytorch_lightning.LightningModule.training_step`` or
+            ``pytorch_lightning.LightningModule.validation_epoch_end`` and the names thus depend on
+            how this dictionary is formatted.
+    """
+
+    def __init__(self, trial: optuna.trial.Trial, monitor: str) -> None:
+        _imports.check()
+
+        if not isinstance(trial.study._storage, _CachedStorage):
+            raise ValueError(
+                ":class:`~optuna.integration.PyTorchLightningPruningCallback`"
+                " supports only :class:`~optuna.storages.RDBStorage`."
+            )
+
+        super().__init__()
+
+        self._trial = trial
+        self.monitor = monitor
+
+    def on_validation_end(self, trainer: Trainer, pl_module: LightningModule) -> None:
+        epoch = pl_module.current_epoch
+
+        current_score = trainer.callback_metrics.get(self.monitor)
+        if current_score is None:
+            message = (
+                "The metric '{}' is not in the evaluation logs for pruning. "
+                "Please make sure you set the correct metric name.".format(self.monitor)
+            )
+            warnings.warn(message)
+            return
+
+        # When multi gpu, on_validation_end function executes multi times.
+        # So Exception is ignored to avoid IntegrityError.
+        distributed_backend = trainer.accelerator_connector.distributed_backend
+        try:
+            self._trial.report(current_score, step=epoch)
+        except sqlalchemy.exc.IntegrityError:
+            if distributed_backend is not None:
+                pass
+            else:
+                raise
+
+        if self._trial.should_prune():
+            trainer.should_stop = True
+            try:
+                self._trial.set_user_attr("pruned", True)
+                self._trial.set_user_attr("epoch", epoch)
+            except sqlalchemy.exc.IntegrityError:
+                if distributed_backend is not None:
+                    pass
+                else:
+                    raise
+
+    # Because on_validation_end is executed in spawned processes, when on_fit_end is executed,
+    # it is necessary to report an objective function value for a given step via RDB.
+    def on_fit_end(self, trainer: Trainer, pl_module: LightningModule) -> None:
+        _trial_id = self._trial._trial_id
+        _study = self._trial.study
+        _trial = _study._storage._backend.get_trial(_trial_id)  # type: ignore
+        is_pruned = _trial.user_attrs.get("pruned")
+        epoch = _trial.user_attrs.get("epoch")
+        intermediate_values = _trial.intermediate_values
+        for step, value in intermediate_values.items():
+            self._trial.report(value, step=step)
+
+        if is_pruned:
+            message = "Trial was pruned at epoch {}.".format(epoch)
+            raise optuna.TrialPruned(message)
```

Comment:
if we support non-ddp training code, 

```suggestion
    def on_fit_end(self, trainer: Trainer, pl_module: LightningModule) -> None:

        _trial_id = self._trial._trial_id
        _study = self._trial.study
        is_ddp_backend = trainer.accelerator_connector.distributed_backend is not None

        # in-memory and non-ddp training
        if not isinstance(_study._storage, _CachedStorage) and not is_ddp_backend:
            _trial = _study._storage.get_trial(_trial_id)
        # rdb storage
        elif isinstance(_study._storage, _CachedStorage):
            _trial = _study._storage._backend.get_trial(_trial_id)  # type: ignore
            # rdb storage and ddp training
            # Because on_validation_end is executed in spawned processes, when on_fit_end is executed,
            # it is necessary to report an objective function value for a given step via RDB.
            if is_ddp_backend:
                intermediate_values = _trial.intermediate_values
                for step, value in intermediate_values.items():
                    self._trial.report(value, step=step)
        # in-memory and ddp-training
        else:
            raise ValueError

        is_pruned = _trial.system_attrs.get("pruned")
        epoch = _trial.system_attrs.get("epoch")

        if is_pruned:
            message = "Trial was pruned at epoch {}.".format(epoch)
            raise optuna.TrialPruned(message)

```